File: querier/internal/query/symbols.go
package query

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"

	"dev.research4jar/querier/internal/project"
)

type SymbolRequest struct {
	Command string `json:"command"`
	Arg     string `json:"arg"`
	Direct  bool   `json:"direct,omitempty"`
}

type SymbolResult struct {
	FQN               string          `json:"fqn"`
	SourceJar         string          `json:"source_jar"`
	Attributes        json.RawMessage `json:"attributes"`
	MatchedAnnotation string          `json:"matched_annotation,omitempty"`
}

type SymbolResponse struct {
	Query    SymbolRequest  `json:"query"`
	Results  []SymbolResult `json:"results"`
	Total    int            `json:"total"`
	Page     int            `json:"page"`
	PageSize int            `json:"page_size"`
	Coverage Coverage       `json:"coverage"`
}

// Transitive closure over declared supertypes: the seed FQN expands through
// super_fqn edges and class_interfaces edges, both stored as symbolic
// references, so the walk crosses jar boundaries inside the merged session.
const transitiveImplCTE = `
WITH RECURSIVE impl(fqn) AS (
  VALUES(?)
  UNION
  SELECT c.fqn FROM classes c JOIN impl ON c.super_fqn = impl.fqn
  UNION
  SELECT c.fqn FROM classes c
    JOIN class_interfaces ci ON ci.class_id = c.id
    JOIN impl ON ci.interface_fqn = impl.fqn
)`

// Meta-annotation closure: an annotation type (kind='annotation') annotated by
// a member of the set joins the set. Plain UNION deduplicates, which also
// terminates self-referential annotations such as @Documented.
const metaAnnotationCTE = `
WITH RECURSIVE meta(fqn) AS (
  VALUES(?)
  UNION
  SELECT c.fqn
  FROM classes c
  JOIN annotations a ON a.target_kind = 'class' AND a.target_id = c.id
  JOIN meta ON a.annotation_fqn = meta.fqn
  WHERE c.kind = 'annotation'
)`

func FindImplementations(
	ctx context.Context,
	pointer project.Pointer,
	manifestPath string,
	targetFQN string,
	direct bool,
	page int,
	pageSize int,
) (SymbolResponse, error) {
	countSQL := transitiveImplCTE + `
		SELECT COUNT(*) FROM classes c
		WHERE c.fqn IN (SELECT fqn FROM impl) AND c.fqn <> ?`
	selectSQL := transitiveImplCTE + `
		SELECT c.fqn, c.source_shard_id
		FROM classes c
		WHERE c.fqn IN (SELECT fqn FROM impl) AND c.fqn <> ?
		ORDER BY c.fqn, c.source_shard_id
		LIMIT ? OFFSET ?`
	if direct {
		countSQL = `SELECT COUNT(*) FROM (
			SELECT class_id FROM class_interfaces WHERE interface_fqn = ?
			UNION
			SELECT id FROM classes WHERE super_fqn = ?
		)`
		selectSQL = `SELECT c.fqn, c.source_shard_id
			FROM classes c
			WHERE c.id IN (
			  SELECT class_id FROM class_interfaces WHERE interface_fqn = ?
			  UNION
			  SELECT id FROM classes WHERE super_fqn = ?
			)
			ORDER BY c.fqn, c.source_shard_id
			LIMIT ? OFFSET ?`
	}
	return findSymbols(
		ctx,
		pointer,
		manifestPath,
		SymbolRequest{Command: "find-implementations", Arg: targetFQN, Direct: direct},
		[]any{targetFQN, targetFQN},
		page,
		pageSize,
		countSQL,
		selectSQL,
		func(rows *sql.Rows) (SymbolResult, string, error) {
			var result SymbolResult
			var shardID string
			err := rows.Scan(&result.FQN, &shardID)
			return result, shardID, err
		},
	)
}

func FindByAnnotation(
	ctx context.Context,
	pointer project.Pointer,
	manifestPath string,
	annotationFQN string,
	direct bool,
	page int,
	pageSize int,
) (SymbolResponse, error) {
	countSQL := metaAnnotationCTE + `
		SELECT COUNT(*)
		FROM annotations a
		JOIN classes c ON c.id = a.target_id
		WHERE a.target_kind = 'class' AND a.annotation_fqn IN (SELECT fqn FROM meta)`
	selectSQL := metaAnnotationCTE + `
		SELECT c.fqn, c.source_shard_id, a.attributes, a.annotation_fqn
		FROM annotations a
		JOIN classes c ON c.id = a.target_id
		WHERE a.target_kind = 'class' AND a.annotation_fqn IN (SELECT fqn FROM meta)
		ORDER BY c.fqn, c.source_shard_id, a.annotation_fqn, COALESCE(a.attributes, '')
		LIMIT ? OFFSET ?`
	if direct {
		countSQL = `SELECT COUNT(*)
			FROM annotations a
			JOIN classes c ON c.id = a.target_id
			WHERE a.target_kind = 'class' AND a.annotation_fqn = ?`
		selectSQL = `SELECT c.fqn, c.source_shard_id, a.attributes, a.annotation_fqn
			FROM annotations a
			JOIN classes c ON c.id = a.target_id
			WHERE a.target_kind = 'class' AND a.annotation_fqn = ?
			ORDER BY c.fqn, c.source_shard_id
			LIMIT ? OFFSET ?`
	}
	return findSymbols(
		ctx,
		pointer,
		manifestPath,
		SymbolRequest{Command: "find-by-annotation", Arg: annotationFQN, Direct: direct},
		[]any{annotationFQN},
		page,
		pageSize,
		countSQL,
		selectSQL,
		func(rows *sql.Rows) (SymbolResult, string, error) {
			var result SymbolResult
			var shardID string
			var attributes sql.NullString
			if err := rows.Scan(
				&result.FQN, &shardID, &attributes, &result.MatchedAnnotation,
			); err != nil {
				return result, shardID, err
			}
			if attributes.Valid && json.Valid([]byte(attributes.String)) {
				result.Attributes = json.RawMessage(attributes.String)
			}
			return result, shardID, nil
		},
	)
}

func findSymbols(
	ctx context.Context,
	pointer project.Pointer,
	manifestPath string,
	request SymbolRequest,
	bindArgs []any,
	page int,
	pageSize int,
	countSQL string,
	selectSQL string,
	scan func(rows *sql.Rows) (SymbolResult, string, error),
) (SymbolResponse, error) {
	session, err := openReadOnly(pointer.SessionDBPath, true)
	if err != nil {
		return SymbolResponse{}, fmt.Errorf("open session database: %w", err)
	}
	defer session.Close()

	var total int
	if err := session.QueryRowContext(ctx, countSQL, bindArgs...).Scan(&total); err != nil {
		return SymbolResponse{}, fmt.Errorf("count %s results: %w", request.Command, err)
	}
	selectArgs := append(append([]any{}, bindArgs...), pageSize, (page-1)*pageSize)
	rows, err := session.QueryContext(ctx, selectSQL, selectArgs...)
	if err != nil {
		return SymbolResponse{}, fmt.Errorf("query %s results: %w", request.Command, err)
	}
	defer rows.Close()

	type pendingResult struct {
		result  SymbolResult
		shardID string
	}
	pending := make([]pendingResult, 0)
	for rows.Next() {
		result, shardID, err := scan(rows)
		if err != nil {
			return SymbolResponse{}, fmt.Errorf("scan %s result: %w", request.Command, err)
		}
		pending = append(pending, pendingResult{result: result, shardID: shardID})
	}
	if err := rows.Err(); err != nil {
		return SymbolResponse{}, fmt.Errorf("iterate %s results: %w", request.Command, err)
	}

	var sources map[string]string
	if len(pending) > 0 {
		sources, err = loadSourceJars(ctx, manifestPath)
		if err != nil {
			return SymbolResponse{}, err
		}
	}
	results := make([]SymbolResult, 0, len(pending))
	for _, item := range pending {
		item.result.SourceJar = sources[item.shardID]
		if item.result.SourceJar == "" {
			item.result.SourceJar = item.shardID
		}
		results = append(results, item.result)
	}

	return SymbolResponse{
		Query:    request,
		Results:  results,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
		Coverage: coverageFrom(pointer),
	}, nil
}

func coverageFrom(pointer project.Pointer) Coverage {
	return Coverage{
		JarsTotal:        pointer.Coverage.JarsTotal,
		JarsIndexed:      pointer.Coverage.JarsIndexed,
		JarsMissing:      nonNil(pointer.Coverage.JarsMissing),
		ExtractorVersion: pointer.ExtractorVersion,
	}
}

File: core/src/main/kotlin/dev/research4jar/query/Symbols.kt
package dev.research4jar.query

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonRawValue
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.sql.ResultSet

data class SymbolResult(
    @JsonProperty("fqn") val fqn: String,
    @JsonProperty("source_jar") val sourceJar: String,
    @JsonRawValue
    @JsonProperty("attributes") val attributes: String?,
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonProperty("matched_annotation") val matchedAnnotation: String = "",
)

data class SymbolResponse(
    @JsonProperty("query") val query: SymbolRequest,
    @JsonProperty("results") val results: List<SymbolResult>,
    @JsonProperty("total") val total: Int,
    @JsonProperty("page") val page: Int,
    @JsonProperty("page_size") val pageSize: Int,
    @JsonProperty("coverage") val coverage: Coverage,
)

// Transitive closure over declared supertypes: the seed FQN expands through
// super_fqn edges and class_interfaces edges, both stored as symbolic
// references, so the walk crosses jar boundaries inside the merged session.
private const val TRANSITIVE_IMPL_CTE = """
WITH RECURSIVE impl(fqn) AS (
  VALUES(?)
  UNION
  SELECT c.fqn FROM classes c JOIN impl ON c.super_fqn = impl.fqn
  UNION
  SELECT c.fqn FROM classes c
    JOIN class_interfaces ci ON ci.class_id = c.id
    JOIN impl ON ci.interface_fqn = impl.fqn
)"""

// Meta-annotation closure: an annotation type (kind='annotation') annotated by
// a member of the set joins the set. Plain UNION deduplicates, which also
// terminates self-referential annotations such as @Documented.
private const val META_ANNOTATION_CTE = """
WITH RECURSIVE meta(fqn) AS (
  VALUES(?)
  UNION
  SELECT c.fqn
  FROM classes c
  JOIN annotations a ON a.target_kind = 'class' AND a.target_id = c.id
  JOIN meta ON a.annotation_fqn = meta.fqn
  WHERE c.kind = 'annotation'
)"""

private val rawJsonMapper = jacksonObjectMapper()

private fun validRawJson(text: String?): String? {
    if (text == null) return null
    return try {
        rawJsonMapper.readTree(text)
        text
    } catch (_: com.fasterxml.jackson.core.JacksonException) {
        null
    }
}

fun findImplementations(
    pointer: ProjectPointerData,
    manifestPath: String,
    targetFqn: String,
    direct: Boolean,
    page: Int,
    pageSize: Int,
): SymbolResponse {
    var countSql = TRANSITIVE_IMPL_CTE + """
        SELECT COUNT(*) FROM classes c
        WHERE c.fqn IN (SELECT fqn FROM impl) AND c.fqn <> ?"""
    var selectSql = TRANSITIVE_IMPL_CTE + """
        SELECT c.fqn, c.source_shard_id
        FROM classes c
        WHERE c.fqn IN (SELECT fqn FROM impl) AND c.fqn <> ?
        ORDER BY c.fqn, c.source_shard_id
        LIMIT ? OFFSET ?"""
    if (direct) {
        countSql = """SELECT COUNT(*) FROM (
            SELECT class_id FROM class_interfaces WHERE interface_fqn = ?
            UNION
            SELECT id FROM classes WHERE super_fqn = ?
        )"""
        selectSql = """SELECT c.fqn, c.source_shard_id
            FROM classes c
            WHERE c.id IN (
              SELECT class_id FROM class_interfaces WHERE interface_fqn = ?
              UNION
              SELECT id FROM classes WHERE super_fqn = ?
            )
            ORDER BY c.fqn, c.source_shard_id
            LIMIT ? OFFSET ?"""
    }
    return findSymbols(
        pointer,
        manifestPath,
        SymbolRequest(command = "find-implementations", arg = targetFqn, direct = direct),
        listOf(targetFqn, targetFqn),
        page,
        pageSize,
        countSql,
        selectSql,
    ) { rows ->
        SymbolResult(
            fqn = rows.getString(1),
            sourceJar = "",
            attributes = null,
        ) to rows.getString(2)
    }
}

fun findByAnnotation(
    pointer: ProjectPointerData,
    manifestPath: String,
    annotationFqn: String,
    direct: Boolean,
    page: Int,
    pageSize: Int,
): SymbolResponse {
    var countSql = META_ANNOTATION_CTE + """
        SELECT COUNT(*)
        FROM annotations a
        JOIN classes c ON c.id = a.target_id
        WHERE a.target_kind = 'class' AND a.annotation_fqn IN (SELECT fqn FROM meta)"""
    var selectSql = META_ANNOTATION_CTE + """
        SELECT c.fqn, c.source_shard_id, a.attributes, a.annotation_fqn
        FROM annotations a
        JOIN classes c ON c.id = a.target_id
        WHERE a.target_kind = 'class' AND a.annotation_fqn IN (SELECT fqn FROM meta)
        ORDER BY c.fqn, c.source_shard_id, a.annotation_fqn, COALESCE(a.attributes, '')
        LIMIT ? OFFSET ?"""
    if (direct) {
        countSql = """SELECT COUNT(*)
            FROM annotations a
            JOIN classes c ON c.id = a.target_id
            WHERE a.target_kind = 'class' AND a.annotation_fqn = ?"""
        selectSql = """SELECT c.fqn, c.source_shard_id, a.attributes, a.annotation_fqn
            FROM annotations a
            JOIN classes c ON c.id = a.target_id
            WHERE a.target_kind = 'class' AND a.annotation_fqn = ?
            ORDER BY c.fqn, c.source_shard_id, a.annotation_fqn, COALESCE(a.attributes, '')
            LIMIT ? OFFSET ?"""
    }
    return findSymbols(
        pointer,
        manifestPath,
        SymbolRequest(command = "find-by-annotation", arg = annotationFqn, direct = direct),
        listOf(annotationFqn),
        page,
        pageSize,
        countSql,
        selectSql,
    ) { rows ->
        SymbolResult(
            fqn = rows.getString(1),
            sourceJar = "",
            attributes = validRawJson(rows.getString(3)),
            matchedAnnotation = rows.getString(4),
        ) to rows.getString(2)
    }
}

private fun findSymbols(
    pointer: ProjectPointerData,
    manifestPath: String,
    request: SymbolRequest,
    bindArgs: List<Any?>,
    page: Int,
    pageSize: Int,
    countSql: String,
    selectSql: String,
    scan: (ResultSet) -> Pair<SymbolResult, String>,
): SymbolResponse = Db.openReadOnly(pointer.sessionDbPath, immutable = true).use { session ->
    val window = pageWindow(page, pageSize)
    val total = session.queryInt(countSql, bindArgs)
    val pending = session.query(
        selectSql,
        bindArgs + listOf(window.limit, window.offset),
    ) { rows -> rows.mapRows(scan) }

    val sources = if (pending.isNotEmpty()) {
        ManifestCache.loadSourceJars(session, manifestPath, pending.map { it.second })
    } else {
        null
    }
    SymbolResponse(
        query = request,
        results = pending.map { (result, shardId) ->
            result.copy(sourceJar = sourceJarName(sources, shardId))
        },
        total = total,
        page = page,
        pageSize = pageSize,
        coverage = coverageFrom(pointer),
    )
}
